cvnp/cvnp.cpp

```diff
@@ -9,7 +9,59 @@ namespace cvnp
     namespace detail
     {
         namespace py = pybind11;
-        
+
+        // Translated from cv2_numpy.cpp in OpenCV source code
+        class CvnpAllocator : public cv::MatAllocator {
+        public:
+            CvnpAllocator() { stdAllocator = cv::Mat::getStdAllocator(); }
+            ~CvnpAllocator() {}
+
+            cv::UMatData* allocate(pybind11::array& a) const {
+                //int dims, const int* sizes, int type, size_t* step) const {
+                cv::UMatData* u = new cv::UMatData(this);
+                u->data = u->origdata = (uchar*)a.mutable_data(0);
+                // auto *strides = a.strides();
+                // for( int i = 0; i < dims - 1; i++ )
+                //     step[i] = (size_t)_strides[i];
+                // step[dims-1] = CV_ELEM_SIZE(type);
+                // u->size = sizes[0]*step[0];
+                u->size = a.size();
+                u->userdata = a.inc_ref().ptr();
+                return u;
+            }
+
+            cv::UMatData* allocate(int dims0, const int* sizes, int type, void* data, size_t* step, cv::AccessFlag flags, cv::UMatUsageFlags usageFlags) const override
+            {
+                // TODO: I don't think this is ever called
+                throw py::value_error("should never happen");
+                // return stdAllocator->allocate(dims0, sizes, type, data, step, flags, usageFlags);
+            }
+
+            bool allocate(cv::UMatData* u, cv::AccessFlag accessFlags, cv::UMatUsageFlags usageFlags) const override {
+                // TODO: I don't think this is ever called
+                throw py::value_error("should never happen");
+                // return stdAllocator->allocate(u, accessFlags, usageFlags);
+            }
+
+            void deallocate(cv::UMatData* u) const override {
+                if(!u)
+                    return;
+
+                py::gil_scoped_acquire gil;
+                assert(u->urefcount >= 0);
+                assert(u->refcount >= 0);
+                if(u->refcount == 0)
+                {
+                    PyObject* o = (PyObject*)u->userdata;
+                    Py_XDECREF(o);
+                    delete u;
+                }
```

Comment:
Why are there two different ways to handle the number of references for the array?

* a simple `a.inc_ref()` inside `allocate`
* And here, we have to deal with the GIL

I'm sure you tried the simpler version before, but I'd like to know which issue you encountered. I'm always a bit concerned when I see that the GIL is concerned.

One reassuring note: your code did work inside the bindings of ImGui Bundle, and it means that it is compatible with the use of multiple threads that will use each in turn the GIL (I had to implement that when adding a tool called "ImGui Test Engine" to it)

Reply:
There is an implicit assumption that when you call allocate that you have the GIL -- which maybe that should be documented?

Here, we could be called anywhere that OpenCV brings the reference count to zero, so we have to ensure we have the GIL. 

Using the raw C API is easier here because the thing stored in userdata is a PyObject.

Reply:
a thread switch might perhaps occur in between the bound function call and the call to inc_ref inside allocate; In which case we should ensure to have the GIL in both cases. 
 
 what do you think ?

Reply:
Allocate is called by nparray_to_mat which is called by the type caster. If the GIL is not held when calling the type caster, then something is terribly wrong.